Add reproducible charm build feature (#585)
juju/charm-tools: docs/reproducible-charms.rst

 * ``--write-lock-file``
 * ``--use-lock-file-branches``
 * ``--ignore-lock-file``

Creating the lock file
----------------------

To create a lock file, the option ``--write-lock-file`` is passed to the
``charm build`` command.  This option *automatically* ignores the existing lock
file, and rebuilds the charm using the latest versions of the layers and the
versions of the modules as determined in the various ``wheelhouse.txt`` files.

Python module versions are also recorded.  If a VCS repository is used for the
python module, then any branch specified is also recorded, along with the
commit.

At the end of the build, the lock file is written with all of the layer and
Python module information.

The lock file is installed *in* the base layer directory so that it can be
committed into the VCS and used for subsequent builds.

The name of the lock file is ``build.lock``.

Rebuilding the charm from the lock file
---------------------------------------

If a lock file (``build.lock``) is available in the top layer, then it will be
used to control the versions of the layers and modules *by default*.  i.e. the
presence of the lock file controls the build.

> Two options are available to control the build when a lock file is present:

Reviewer: ```suggestion
Three options are available which can influence the build when a lock file is present:
```

Author: I'll revise the doc as per the suggestion above.  The 'good' news with the patch is that it only comes into being if the relevant lock file is present; otherwise it has no effect.  So this should mitigate any issues around the extra feature.  Also, I've ensured that the existing tests still pass, so there should be no regressions.  I'll do the fix, update, and drop the WIP.  thanks!
